From 7a5382abb63be58554d5087eccf72d891346d417 Mon Sep 17 00:00:00 2001
From: Aishwarya Mallampati <amallampati@google.com>
Date: Thu, 7 May 2026 10:50:54 -0700
Subject: [PATCH 1/3] Restrict USSD requests to the subscription's associated
 user.

Add a permission check in handleUssdRequest to ensure that the calling user is associated with the provided subscription ID.

Bug: 471796626
Test: atest
Flag: EXEMPT BUGFIX
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:b76670f3c5ba3d8ad156de232aff09289fe49e26
Merged-In: I7c24f69688b36e815972b0269f88e1a40059238b
Change-Id: I7c24f69688b36e815972b0269f88e1a40059238b
---
 .../android/phone/PhoneInterfaceManager.java  |  5 ++
 .../phone/PhoneInterfaceManagerTest.java      | 60 ++++++++++++++++---
 2 files changed, 57 insertions(+), 8 deletions(-)

diff --git a/src/com/android/phone/PhoneInterfaceManager.java b/src/com/android/phone/PhoneInterfaceManager.java
index 2366e05..58f5a3e 100644
--- a/src/com/android/phone/PhoneInterfaceManager.java
+++ b/src/com/android/phone/PhoneInterfaceManager.java
@@ -3194,6 +3194,11 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
         enforceTelephonyFeatureWithException(getCurrentPackageName(),
                 PackageManager.FEATURE_TELEPHONY_RADIO_ACCESS, "handleUssdRequest");
 
+        if (!TelephonyPermissions.checkSubscriptionAssociatedWithUser(
+                mApp, subId, Binder.getCallingUserHandle())) {
+            throw new SecurityException(
+                    "The subscription " + subId + " is not associated " + "with the calling user");
+        }
         final long identity = Binder.clearCallingIdentity();
         try {
             if (!SubscriptionManager.isValidSubscriptionId(subId)) {
diff --git a/tests/src/com/android/phone/PhoneInterfaceManagerTest.java b/tests/src/com/android/phone/PhoneInterfaceManagerTest.java
index c819a53..0e93f72 100644
--- a/tests/src/com/android/phone/PhoneInterfaceManagerTest.java
+++ b/tests/src/com/android/phone/PhoneInterfaceManagerTest.java
@@ -75,6 +75,7 @@ import android.telephony.NetworkSecurityEvent;
 import android.telephony.RadioAccessFamily;
 import android.telephony.Rlog;
 import android.telephony.SubscriptionInfo;
+import android.telephony.SubscriptionManager;
 import android.telephony.TelephonyManager;
 import android.telephony.UiccPortInfo;
 import android.telephony.UiccSlotInfo;
@@ -139,8 +140,6 @@ public class PhoneInterfaceManagerTest extends TelephonyTestBase {
     private static final String sDebugPackageName =
             PhoneInterfaceManagerTest.class.getPackageName();
 
-    @Mock
-    Phone mPhone;
     @Mock
     FeatureFlags mFeatureFlags;
     @Mock
@@ -148,15 +147,11 @@ public class PhoneInterfaceManagerTest extends TelephonyTestBase {
     @Mock
     private SubscriptionManagerService mSubscriptionManagerService;
     @Mock
-    private com.android.internal.telephony.data.DataNetworkController mDataNetworkController;
-
-    @Mock
-    private AppOpsManager mAppOps;
-    @Mock
     private android.media.AudioManager mAudioManager;
     @Mock
     private SatelliteController mSatelliteController;
-
+    @Mock
+    private AppOpsManager mAppOps;
     @Rule public final SetFlagsRule mSetFlagsRule = new SetFlagsRule();
     private UiccController mUiccController = null;
     private PinStorage mPinStorage = null;
@@ -206,6 +201,9 @@ public class PhoneInterfaceManagerTest extends TelephonyTestBase {
         doReturn(mSubscriptionManagerService).when(mPhoneInterfaceManager)
                 .getSubscriptionManagerService();
         TelephonyManager.setupISubForTest(mSubscriptionManagerService);
+        replaceInstance(SubscriptionManagerService.class, "sInstance", null,
+                mSubscriptionManagerService);
+        doReturn(new int[0]).when(mSubscriptionManagerService).getActiveSubIdList(anyBoolean());
 
         // Some message handlers query these methods on the default phone instance.
         // Make sure they return sensible values and the mPhone mock instance is set
@@ -1330,4 +1328,50 @@ public class PhoneInterfaceManagerTest extends TelephonyTestBase {
         verify(mSatelliteController, never()).requestSatelliteEnabled(anyBoolean(),
                 anyBoolean(), anyBoolean(), any());
     }
+
+    @Test
+    public void testHandleUssdRequest_SubscriptionNotAssociatedWithUser_ThrowsSecurityException() {
+        int subId = 1;
+        String ussdRequest = "*121#";
+        ResultReceiver callback = mock(ResultReceiver.class);
+
+        // Mock subscription NOT associated with user
+        doReturn(false)
+                .when(mSubscriptionManagerService)
+                .isSubscriptionAssociatedWithUser(eq(subId), any(UserHandle.class));
+
+        assertThrows(
+                SecurityException.class,
+                () -> mPhoneInterfaceManager.handleUssdRequest(subId, ussdRequest, callback));
+    }
+
+    @Test
+    public void testHandleUssdRequest_ValidSubscription_Success() {
+        int subId = 1;
+        String ussdRequest = "*121#";
+        ResultReceiver callback = mock(ResultReceiver.class);
+
+        // Mock subscription IS associated with user
+        doReturn(true)
+                .when(mSubscriptionManagerService)
+                .isSubscriptionAssociatedWithUser(eq(subId), any(UserHandle.class));
+        SubscriptionManager sm = (SubscriptionManager) mContext.getSystemService(
+                Context.TELEPHONY_SUBSCRIPTION_SERVICE);
+        doReturn(true).when(sm).isSubscriptionAssociatedWithUser(eq(subId), any(UserHandle.class));
+
+        doReturn(true)
+                .when(mPackageManager)
+                .hasSystemFeature(PackageManager.FEATURE_TELEPHONY_RADIO_ACCESS);
+
+        // Should not throw SecurityException. It will throw RuntimeException because of deadlock
+        // since we are calling it from the main thread. Reaching that point means the security
+        // check passed.
+        try {
+            mPhoneInterfaceManager.handleUssdRequest(subId, ussdRequest, callback);
+        } catch (RuntimeException e) {
+            if (!e.getMessage().contains("deadlock")) {
+                throw e;
+            }
+        }
+    }
 }
-- 
2.55.0

